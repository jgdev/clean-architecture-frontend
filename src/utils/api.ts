import { reactive, unref } from "vue";
import { FetchAction, PaginatedParams, PaginatedResult } from "../types";
import { authSession } from "../api";

const apiUrl = import.meta.env.VITE_API_BASE_URL;

export const createFetchAction = <T, K = any>(
  apiEndpoint: string,
  defaultValue?: T,
  transform?: {
    response?: (body: any) => any;
  }
): FetchAction<T, K> => {
  const state = reactive({
    result: defaultValue,
    error: undefined,
  }) as FetchAction<T, K>;

  const fetchAction = (defaultOptions: RequestInit) => {
    return async (
      options?: RequestInit & {
        data?: K;
      }
    ): Promise<void> => {
      const requestOptions: RequestInit = {
        ...defaultOptions,
        ...options,
      };

      requestOptions.headers = {
        ...(requestOptions.headers || {}),
        ...(authSession.result && {
          "X-Session-Id": authSession.result,
        }),
      };

      state.loading = true;

      return fetch(apiUrl + apiEndpoint, {
        ...requestOptions,
        ...(requestOptions?.method !== "GET" &&
          options?.data && {
            body: JSON.stringify(options.data),
            headers: {
              ...(requestOptions.headers || {}),
              "Content-Type": "application/json",
            },
          }),
      })
        .then(async (res) => {
          const body = await res.json();
          if (!res.ok) return Promise.reject(body);
          return body as K;
        })
        .then(({ result }: any) => {
          state.result =
            (transform?.response && transform.response(result)) || result;
          return result;
        })
        .catch((err) => {
          state.error = err;
          return Promise.reject(err);
        })
        .finally(() => {
          state.loading = false;
        });
    };
  };

  state.delAction = fetchAction({ method: "DEL" });
  state.postAction = fetchAction({ method: "POST" });
  state.getAction = fetchAction({ method: "GET" });

  return unref(state);
};

export const createFetchResourcesAction = <T>(
  apiEndpoint: string,
  transform?: {
    response?: (body: any) => any;
  }
): FetchAction<PaginatedResult<T>> => {
  const state = createFetchAction<PaginatedResult<T>, PaginatedParams>(
    apiEndpoint,
    {
      result: [],
      limit: 0,
      searchParams: undefined,
      skip: 0,
    },
    transform
  );

  return state;
};
