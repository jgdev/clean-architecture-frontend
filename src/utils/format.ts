export const formatAmount = (amount: number) => {
  return amount.toLocaleString("en-US", {
    style: "currency",
    currency: "USD",
    currencyDisplay: "symbol",
  });
};

export const limitCharacters = (str: string, limit: number) => {
  return str.length > limit ? str.substring(0, limit) + '...' : str
}