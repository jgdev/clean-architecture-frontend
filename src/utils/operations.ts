export const stringGeneratorOptions = [
  ["Alphabetic", true],
  ["Uppercase", true],
  ["Lowercase", true],
  ["Numbers", true],
  ["Symbols", true],
];

export const mapConfiguration: { [key: string]: any } = {
  addition: {
    value: "addition",
    name: "Addition",
    input: {
      type: "number",
      multiple: true,
      minInput: 2,
      getLabel: (index: number) => (index === 0 ? "This number" : "Plus"),
      getPlaceholder: (index: number) => `Example: ${index + 1}`,
    },
  },
  subtraction: {
    value: "subtraction",
    name: "Subtract",
    input: {
      type: "number",
      multiple: true,
      minInput: 2,
      getLabel: (index: number) =>
        index === 0 ? "This number" : "Substracting",
      getPlaceholder: (index: number) => `Example: ${index + 1}`,
    },
  },
  multiplication: {
    value: "multiplication",
    name: "Multiply",
    input: {
      type: "number",
      multiple: true,
      minInput: 2,
      getLabel: (index: number) =>
        index === 0 ? "This number" : "Multiplied by",
      getPlaceholder: (index: number) => `Example: ${index + 1}`,
    },
  },
  division: {
    value: "division",
    name: "Divide",
    input: {
      type: "number",
      multiple: true,
      minInput: 2,
      getLabel: (index: number) => (index === 0 ? "This number" : "Divided by"),
      getPlaceholder: (index: number) => `Example: ${index + 1}`,
    },
  },
  square_root: {
    value: "square_root",
    name: "Square root",
    input: {
      type: "number",
      multiple: false,
      getLabel: () => "Square root of ...",
      getPlaceholder: () => "Type the base number",
    },
  },
  random_string: {
    value: "random_string",
    name: "Random String",
    input: {
      type: "number",
      multiple: false,
      options: stringGeneratorOptions,
      getLabel: () => "String length",
      getPlaceholder: () => "Example: 10",
    },
  },
  random_string_v2: {
    value: "random_string_v2",
    name: "Random String (Random.org)",
    input: {
      type: "number",
      multiple: false,
      options: stringGeneratorOptions,
      maxLength: 32,
      getLabel: () => "String length",
      getPlaceholder: () => "Example: 10",
    },
  },
};
