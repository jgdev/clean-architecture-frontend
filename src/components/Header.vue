<script lang="ts" setup>
import { Button, Spinner } from 'flowbite-vue';
import { user, authLogout, authSession, records } from '../api';
import { formatAmount } from '../utils/format';

type Props = {
  onPerformOperation: () => void;
}

defineProps<Props>()

const handleLogout = (e: Event) => {
  e.preventDefault();
  authLogout.postAction().then(() => {
    authSession.result = ''
  })
}
</script>

<template>
  <div class="sm:top-auto sm:relative sm:border-b p-4 pr-4 sm:p-6 bg-gray-100 w-full border sm:rounded-lg main-header">
    <div class="sm:max-w-2xl sm:m-auto">
      <div class="w-full flex justify-between items-center border-b mb-4 pb-4 text-gray-600">
      <h1 class="text-sm">Arithmetic App</h1>
      <a href="#" @click="handleLogout" class="text-sm font-bold">Logout</a>
    </div>
    <div class="w-full flex justify-between items-center gap-2">
      <Button size="sm" color="light" class="shadow" :disabled="records.loading || user.loading"
        @click="() => onPerformOperation()">Perform
        operation</Button>
      <div class="flex flex-col balance items-center justify-center">
        <span class="text-xs font-bold" v-if="!user.loading">User balance</span>
        <div class="text-right w-full text-xs">
          <Spinner size="4" color="blue" v-if="!!user.loading" />
          {{ !user.loading && formatAmount(user.result?.balance || 0) || '' }}
        </div>
      </div>
    </div>
    </div>
  </div>
</template>

<style scoped>
.main-header {
  max-height: 8rem;
}
</style>