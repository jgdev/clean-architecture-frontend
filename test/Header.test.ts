import { mount } from "@vue/test-utils";
import { createFakeAction, createFakeResources } from "./utils";

import {
  FetchAction,
  FetchResources,
  Operation,
  Record,
  User,
} from "../src/types";

import Header from "../src/components/Header.vue";

describe("Header", () => {
  let user: FetchAction<User>;
  let authSession: FetchAction<string>;
  let authLogout: FetchAction<void>;
  let records: FetchResources<Record>;
  let operations: FetchResources<Operation>;
  let deleteRecord: FetchAction<void>;
  let onPerformOperation = vi.fn();

  beforeEach(() => {
    onPerformOperation = vi.fn();
    user = createFakeAction<User>({
      balance: 200,
      email: "test@test",
    });
    authSession = createFakeAction<string>();
    authLogout = createFakeAction<void>();
    deleteRecord = createFakeAction<void>();
    records = createFakeResources<Record>();
    operations = createFakeResources<Operation>();
  });

  it("should display user balance properly", () => {
    const wrapper = mount(Header, {
      props: {
        user,
        authLogout,
        authSession,
        records,
        onPerformOperation,
      },
    });
    expect(wrapper.text()).toMatch(/\$200.00/);
    expect(wrapper.find("svg#loading-user-balance").exists()).toBe(false);
    expect(wrapper.find("#user-balance").exists()).toBe(true);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should hide balance and disable controls while loading profile", () => {
    user.loading = true;
    const wrapper = mount(Header, {
      props: {
        user,
        authLogout,
        authSession,
        records,
        onPerformOperation,
      },
    });
    expect(wrapper.find("svg#loading-user-balance").exists()).toBe(true);
    expect(wrapper.find("#user-balance").exists()).toBe(false);
    expect(wrapper.find("button").attributes()).toHaveProperty("disabled");
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should handle perform new operation click", async () => {
    user.loading = false;
    const wrapper = mount(Header, {
      props: {
        user,
        authLogout,
        authSession,
        records,
        onPerformOperation,
      },
    });
    await wrapper.find("button").trigger("click");
    expect(onPerformOperation).toHaveBeenCalledOnce();
  });

  it("should handle logout", async () => {
    user.loading = false;
    const wrapper = mount(Header, {
      props: {
        user,
        authLogout,
        authSession,
        records,
        onPerformOperation,
      },
    });
    await wrapper.find("a").trigger("click");
    expect(authLogout.postAction).toHaveBeenCalledOnce();
  });
});
