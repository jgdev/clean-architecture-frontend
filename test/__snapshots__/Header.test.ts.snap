// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Header > should display user balance properly 1`] = `
"<div data-v-c970699f=\\"\\" class=\\"sm:top-auto sm:relative sm:border-b p-4 pr-4 sm:p-6 bg-gray-100 w-full border main-header\\">
  <div data-v-c970699f=\\"\\" class=\\"sm:max-w-2xl sm:m-auto\\">
    <div data-v-c970699f=\\"\\" class=\\"w-full flex justify-between items-center border-b mb-4 pb-4 text-gray-600\\">
      <h1 data-v-c970699f=\\"\\" class=\\"text-sm\\">Arithmetic App</h1><a data-v-c970699f=\\"\\" href=\\"#\\" class=\\"text-sm font-bold\\">Logout</a>
    </div>
    <div data-v-c970699f=\\"\\" class=\\"w-full flex justify-between items-center gap-2\\"><button data-v-c970699f=\\"\\" class=\\"text-gray-900 bg-white border-gray-300 focus:outline-none focus:ring-4 font-medium rounded-lg dark:bg-gray-800 dark:text-white dark:border-gray-600 dark:focus:ring-gray-700 hover:bg-gray-100 dark:hover:border-gray-600 px-3 py-1.5 text-gray-900 bg-white border border-gray-300 focus:outline-none focus:ring-4 focus:ring-gray-200 font-medium rounded-lg dark:bg-gray-800 dark:text-white dark:border-gray-600 dark:focus:ring-gray-700 hover:bg-gray-100 dark:hover:border-gray-600 text-sm px-3 py-1.5 shadow\\">
        <!----><span class=\\"\\"><!---->Perform operation<!----></span>
        <!---->
      </button>
      <div data-v-c970699f=\\"\\" class=\\"flex flex-col balance items-center justify-center\\"><span data-v-c970699f=\\"\\" class=\\"text-xs font-bold\\">User balance</span>
        <div data-v-c970699f=\\"\\" class=\\"text-right w-full text-xs\\">
          <!--v-if--><span data-v-c970699f=\\"\\" id=\\"user-balance\\">$200.00</span>
        </div>
      </div>
    </div>
  </div>
</div>"
`;

exports[`Header > should hide balance and disable controls while loading profile 1`] = `
"<div data-v-c970699f=\\"\\" class=\\"sm:top-auto sm:relative sm:border-b p-4 pr-4 sm:p-6 bg-gray-100 w-full border main-header\\">
  <div data-v-c970699f=\\"\\" class=\\"sm:max-w-2xl sm:m-auto\\">
    <div data-v-c970699f=\\"\\" class=\\"w-full flex justify-between items-center border-b mb-4 pb-4 text-gray-600\\">
      <h1 data-v-c970699f=\\"\\" class=\\"text-sm\\">Arithmetic App</h1><a data-v-c970699f=\\"\\" href=\\"#\\" class=\\"text-sm font-bold\\">Logout</a>
    </div>
    <div data-v-c970699f=\\"\\" class=\\"w-full flex justify-between items-center gap-2\\"><button data-v-c970699f=\\"\\" class=\\"text-gray-900 bg-white border-gray-300 focus:outline-none focus:ring-4 font-medium rounded-lg dark:bg-gray-800 dark:text-white dark:border-gray-600 dark:focus:ring-gray-700 cursor-not-allowed opacity-50 px-3 py-1.5 text-gray-900 bg-white border border-gray-300 focus:outline-none focus:ring-4 focus:ring-gray-200 font-medium rounded-lg dark:bg-gray-800 dark:text-white dark:border-gray-600 dark:focus:ring-gray-700 cursor-not-allowed opacity-50 text-sm px-3 py-1.5 shadow\\" disabled=\\"\\">
        <!----><span class=\\"\\"><!---->Perform operation<!----></span>
        <!---->
      </button>
      <div data-v-c970699f=\\"\\" class=\\"flex flex-col balance items-center justify-center\\">
        <!--v-if-->
        <div data-v-c970699f=\\"\\" class=\\"text-right w-full text-xs\\"><svg data-v-c970699f=\\"\\" role=\\"status\\" class=\\"w-4 h-4 text-gray-200 dark:text-gray-600 fill-blue-600 animate-spin\\" viewBox=\\"0 0 100 101\\" fill=\\"none\\" xmlns=\\"http://www.w3.org/2000/svg\\" id=\\"loading-user-balance\\">
            <path d=\\"M100 50.5908C100 78.2051 77.6142 100.591 50 100.591C22.3858 100.591 0 78.2051 0 50.5908C0 22.9766 22.3858 0.59082 50 0.59082C77.6142 0.59082 100 22.9766 100 50.5908ZM9.08144 50.5908C9.08144 73.1895 27.4013 91.5094 50 91.5094C72.5987 91.5094 90.9186 73.1895 90.9186 50.5908C90.9186 27.9921 72.5987 9.67226 50 9.67226C27.4013 9.67226 9.08144 27.9921 9.08144 50.5908Z\\" fill=\\"currentColor\\"></path>
            <path d=\\"M93.9676 39.0409C96.393 38.4038 97.8624 35.9116 97.0079 33.5539C95.2932 28.8227 92.871 24.3692 89.8167 20.348C85.8452 15.1192 80.8826 10.7238 75.2124 7.41289C69.5422 4.10194 63.2754 1.94025 56.7698 1.05124C51.7666 0.367541 46.6976 0.446843 41.7345 1.27873C39.2613 1.69328 37.813 4.19778 38.4501 6.62326C39.0873 9.04874 41.5694 10.4717 44.0505 10.1071C47.8511 9.54855 51.7191 9.52689 55.5402 10.0491C60.8642 10.7766 65.9928 12.5457 70.6331 15.2552C75.2735 17.9648 79.3347 21.5619 82.5849 25.841C84.9175 28.9121 86.7997 32.2913 88.1811 35.8758C89.083 38.2158 91.5421 39.6781 93.9676 39.0409Z\\" fill=\\"currentFill\\"></path>
          </svg>
          <!--v-if-->
        </div>
      </div>
    </div>
  </div>
</div>"
`;
